source/extensions/stat_sinks/metrics_service/grpc_metrics_service_impl.h

```diff
@@ -20,79 +20,106 @@ namespace Extensions {
+template <class RequestProto, class ResponseProto>
+class GrpcMetricsStreamer : public Grpc::AsyncStreamCallbacks<ResponseProto> {
 public:
+  GrpcMetricsStreamer(Grpc::AsyncClientFactoryPtr&& factory) : client_(factory->create()) {}
   ~GrpcMetricsStreamer() override = default;
 
   /**
    * Send Metrics Message.
    * @param message supplies the metrics to send.
    */
-  virtual void send(envoy::service::metrics::v3::StreamMetricsMessage& message) PURE;
+  virtual void send(MetricsPtr&& metrics) PURE;
 
   // Grpc::AsyncStreamCallbacks
   void onCreateInitialMetadata(Http::RequestHeaderMap&) override {}
   void onReceiveInitialMetadata(Http::ResponseHeaderMapPtr&&) override {}
-  void
-  onReceiveMessage(std::unique_ptr<envoy::service::metrics::v3::StreamMetricsResponse>&&) override {
-  }
+  void onReceiveMessage(std::unique_ptr<ResponseProto>&&) override {}
   void onReceiveTrailingMetadata(Http::ResponseTrailerMapPtr&&) override {}
   void onRemoteClose(Grpc::Status::GrpcStatus, const std::string&) override{};
+
+protected:
+  Grpc::AsyncStream<RequestProto> stream_{};
+  Grpc::AsyncClient<RequestProto, ResponseProto> client_;
 };
 
-using GrpcMetricsStreamerSharedPtr = std::shared_ptr<GrpcMetricsStreamer>;
+template <class RequestProto, class ResponseProto>
+using GrpcMetricsStreamerSharedPtr =
+    std::shared_ptr<GrpcMetricsStreamer<RequestProto, ResponseProto>>;
 
 /**
  * Production implementation of GrpcMetricsStreamer
  */
-class GrpcMetricsStreamerImpl : public Singleton::Instance, public GrpcMetricsStreamer {
+class GrpcMetricsStreamerImpl
+    : public Singleton::Instance,
+      public GrpcMetricsStreamer<envoy::service::metrics::v3::StreamMetricsMessage,
+                                 envoy::service::metrics::v3::StreamMetricsResponse> {
 public:
   GrpcMetricsStreamerImpl(Grpc::AsyncClientFactoryPtr&& factory,
                           const LocalInfo::LocalInfo& local_info,
                           envoy::config::core::v3::ApiVersion transport_api_version);
 
   // GrpcMetricsStreamer
-  void send(envoy::service::metrics::v3::StreamMetricsMessage& message) override;
+  void send(MetricsPtr&& metrics) override;
 
   // Grpc::AsyncStreamCallbacks
   void onRemoteClose(Grpc::Status::GrpcStatus, const std::string&) override { stream_ = nullptr; }
 
 private:
-  Grpc::AsyncStream<envoy::service::metrics::v3::StreamMetricsMessage> stream_{};
-  Grpc::AsyncClient<envoy::service::metrics::v3::StreamMetricsMessage,
-                    envoy::service::metrics::v3::StreamMetricsResponse>
-      client_;
   const LocalInfo::LocalInfo& local_info_;
   const Protobuf::MethodDescriptor& service_method_;
   const envoy::config::core::v3::ApiVersion transport_api_version_;
 };
 
 using GrpcMetricsStreamerImplPtr = std::unique_ptr<GrpcMetricsStreamerImpl>;
 
+class MetricsFlusher {
+public:
+  MetricsFlusher(const bool report_counters_as_deltas)
```

Comment:
`explicit` and avoid `const` in value params

Reply:
I like to keep const for value params that should not be modified in a function body even if it doesn't mean anything at the API layer. What is your take for not having it?

Reply:
Mostly for consistency, it's actively discouraged for declarations (https://clang.llvm.org/extra/clang-tidy/checks/readability-avoid-const-params-in-decls.html) and so I find allowing it in definitions to be a bit confusing. 

Definitely not that important, so if you feel strongly about it then its probably fine